services/src/datasets/external/pangaea/meta.rs

```diff
@@ -0,0 +1,624 @@
+                Some(wkt.clone())
+            }
+            _ => None,
+        };
+
+        // Generate new tsv header
+        let header = self.meta.tsv_header();
+
+        let output = match wkt {
+            // No or included geometry
+            None => {
+                // Replace with generated csv header
+                if let Some(idx) = input.find('\n') {
+                    input.replace_range(0..idx, header.as_str());
+                }
+                input
+            }
+            // Append default geometry
+            Some(wkt) => {
+                let mut output: String = String::with_capacity(input.len());
+
+                let mut i = input.lines();
+
+                // Discard source header
+                i.next().unwrap();
+
+                // Write output header
+                output.push_str(header.as_str());
+                output.push('\n');
+
+                for line in i {
+                    output.push_str(line);
+                    output.push('\t');
+                    output.push_str(wkt.as_str());
+                    output.push('\n');
+                }
+                output
+            }
+        };
+        output
+    }
+}
+
+#[async_trait]
+impl PreLoadHook for LoadHook {
+    async fn execute(&self) -> Result<(), geoengine_operators::error::Error> {
+        if self.handle.lock().unwrap().is_some() {
+            return Ok(());
+        }
+
+        // Execute request, read source tsv and process it.
+        let tsv = self.process_tsv(
+            self.client
+                .get(&self.url)
+                .send()
+                .await
+                .map_err(|e| geoengine_operators::error::Error::LoadingInfo {
+                    source: Box::new(e),
+                })?
+                .text()
+                .await
+                .map_err(|e| geoengine_operators::error::Error::LoadingInfo {
+                    source: Box::new(e),
+                })?,
+        );
+
+        // Create the mem-file
+        let mut lock = self.handle.lock().unwrap();
+        if lock.is_none() {
+            lock.replace(MemfileHandle::new(
+                self.memfile_name.clone(),
+                tsv.into_bytes(),
+            )?);
+        }
+        Ok(())
+    }
+
+    fn box_clone(&self) -> Box<dyn PreLoadHook> {
+        Box::new(self.clone())
+    }
+}
+
+/// This struct simply holds the name of a generated gdal memfile and releases
+/// the file on drop.
+#[derive(Debug)]
+struct MemfileHandle {
+    memfile_name: String,
+}
+
+impl MemfileHandle {
+    fn new(
+        memfile_name: String,
+        data: Vec<u8>,
+    ) -> Result<MemfileHandle, geoengine_operators::error::Error> {
+        gdal::vsi::create_mem_file(memfile_name.as_str(), data)?;
+        Ok(MemfileHandle { memfile_name })
+    }
+}
+
+impl Drop for MemfileHandle {
+    fn drop(&mut self) {
+        gdal::vsi::unlink_mem_file(self.memfile_name.as_str()).unwrap();
+    }
+}
```

Comment:
How do you guarantee that this isn't dropped before you access the data?
You get the metadata and then create some loading info and discard this(?).

I guess it has to be put into the stream in order to keep this alive.
But then it is no pre-load-hook but something different :thinking:

Reply:
The Memfile is stored within the hook, and the hook is part of the MetaData. I assumed the MetaData to outlive the stream. If this is not the case, we could make the hook return some result and store it along with the stream (or even store the hook itself).